src/coreclr/jit/scev.cpp

```diff
@@ -1009,3 +1099,662 @@ Scev* ScalarEvolutionContext::Simplify(Scev* scev)
+//   rhs    - The RHS of the expression causing the loop to exit.
+//   exitOp - The relop such that when [lhs] [exitOp] [rhs] evaluates to true, the loop exits.
+//
+// Returns:
+//   True if it is possible for the LHS to overflow its computation type without the loop exiting.
+//   False if we were able to prove that it cannot.
+//
+// Remarks:
+//   May return true conservatively.
+//
+bool ScalarEvolutionContext::MayOverflowBeforeExit(ScevAddRec* lhs, Scev* rhs, VNFunc exitOp)
+{
+    int64_t stepCns;
+    if (!lhs->Step->GetConstantValue(m_comp, &stepCns))
+    {
+        // TODO-CQ: With divisibility checks we can likely handle some of
+        // these.
+        return true;
+    }
+
+    // Handle odd cases, where we count in the other direction of the test, and
+    // thus only exit after overflow (or immediately).
+    //
+    // TODO-CQ: One potential pattern we ought to be able to handle for
+    // downwards counting loops with unsigned indices is:
+    // for (uint i = (uint)arr.Length - 1u; i < (uint)arr.Length; i--)
+    // This one overflows, but on overflow it immediately exits.
+    //
+    switch (exitOp)
+    {
+        case VNF_GE:
+        case VNF_GT:
+        case VNF_GE_UN:
+        case VNF_GT_UN:
+            if (stepCns < 0)
+            {
+                return true;
+            }
+
+            break;
+        case VNF_LE:
+        case VNF_LT:
+        case VNF_LE_UN:
+        case VNF_LT_UN:
+            if (stepCns > 0)
+            {
+                return true;
+            }
+
+            break;
+        default:
+            unreached();
+    }
+
+    // A step count of 1/-1 will always exit for "or equal" checks.
+    if ((stepCns == 1) && ((exitOp == VNFunc(GT_GE)) || (exitOp == VNF_GE_UN)))
+    {
+        return false;
+    }
+
+    if ((stepCns == -1) && ((exitOp == VNFunc(GT_LE)) || (exitOp == VNF_LE_UN)))
+    {
+        return false;
+    }
+
+    // Example: If exitOp is GT_GT then it means that we exit when iv >
+    // limitCns. In the worst case the IV ends up at limitCns and then steps
+    // once more. If that still causes us to exit then no overflow is possible.
+    Scev* step = lhs->Step;
+    if ((exitOp == VNF_GE) || (exitOp == VNF_GE_UN))
+    {
+        // Exit on iv >= limitCns, so in worst case we are at limitCns - 1. Include the -1 in the step.
+        Scev* negOne = NewConstant(rhs->Type, -1);
+        step         = NewBinop(ScevOper::Add, step, negOne);
+    }
+    else if ((exitOp == VNF_LE) || (exitOp == VNF_LE_UN))
+    {
+        // Exit on iv <= limitCns, so in worst case we are at limitCns + 1. Include the +1 in the step.
+        Scev* posOne = NewConstant(rhs->Type, 1);
+        step         = NewBinop(ScevOper::Add, step, posOne);
+    }
+
+    Scev* steppedVal      = NewBinop(ScevOper::Add, rhs, step);
+    steppedVal            = Simplify(steppedVal);
+    ValueNum steppedValVN = MaterializeVN(steppedVal);
```

Comment:
`MaterializeVN` does not create any IR, it directly creates the VN from the SCEV. `Materialize` will always create both IR and VN (to be able to attach the VN to the tree). I wasn't totally sure whether or not that's the behavior we want, but we don't really materialize a lot of IR so it seemed like it wouldn't be expensive, and it's nice to try to keep the invariant that IR has proper VNs at this point.

It would be nice to express SCEVs directly using VNs. The main problem to solve there is that VNs are reduced "too much", making it non-trivial to materialize them into IR. OTOH SCEVs have the property that they can always be materialized in a straightforward way in the preheader. Well, at least that's my current belief.

Reply:
Thanks. Looking again it seems clear that VNs are computed first, then (optionally) IR, then the two are attached. Not sure what I thought I saw the first time.